test/sanity/code-smell/update-bundled.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# (c) 2018, Ansible Project
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+import glob
+import json
+import os.path
+import sys
+from distutils.version import LooseVersion
+
+import packaging.specifiers
+
+from ansible.module_utils.urls import open_url
+
+
+def compat_libs(basedir):
+    for filename in glob.glob(os.path.join(basedir, 'lib/ansible/compat/*/__init__.py')):
+        if 'compat/tests' in filename:
+            # compat/tests doesn't bundle any code
+            continue
+        yield filename
+
+    yield os.path.join(basedir, 'lib/ansible/module_utils/distro/__init__.py')
+    yield os.path.join(basedir, 'lib/ansible/module_utils/six/__init__.py')
+    yield os.path.join(basedir, 'lib/ansible/module_utils/compat/ipaddress.py')
+    # backports.ssl_match_hostname should be moved to its own file
+    yield os.path.join(basedir, 'lib/ansible/module_utils/urls.py')
+
+
+def get_bundled_metadata(filename):
+    with open(filename, 'r') as module:
+        for line in module:
+            if line.strip().startswith('_BUNDLED_METADATA'):
+                data = line[line.index('{'):].strip()
+                break
+        else:
+            raise ValueError('Bundled file {0} without metadata.'
+                             ' Unable to check for update'.format(filename))
+        metadata = json.loads(data)
+    return metadata
+
+
+def get_latest_applicable_version(pypi_data, constraints=None):
+    latest_version = "0"
+    if 'version_constraints' in metadata:
+        version_specification = packaging.specifiers.SpecifierSet(metadata['version_constraints'])
+        for version in pypi_data['releases']:
+            if version in version_specification:
+                if LooseVersion(version) > LooseVersion(latest_version):
+                    latest_version = version
+    else:
+        latest_version = pypi_data['info']['version']
+
+    return latest_version
+
+
+if __name__ == '__main__':
+    basedir = os.path.join(os.path.dirname(__file__), '../../..')
+
+    for filename in compat_libs(basedir):
```

Comment:
Instead of hard-coding a list of files to check, why not check all `*.py` files for `_BUNDLED_METADATA` instead?

That will be easier to maintain and won't require updates to this test every time we add/remove a bundled file.

Reply:
Well, all `*.py` files under any directories where we allow bundling anyways.

Reply:
I thought about that and I'm not sure.  The problem is a social problem and I can't figure out what technical implementations will be better or worse....

What we need is for all bundled code to be checked to make sure that it is up to date.  The problem is how do we detect that there is bundled code?

If we solely key off of the _BUNDLED_METADATA variable being present, then the test will stop detecting that a file needs to be updated if someone drops a new version in place and forgets that they have to add _BUNDLED_METADATA to the new file.

If we operate on files listed in this test case, then someone bundling new code needs to know that they have to add both _BUNDLED_METADATA to their file and list it here.

I think that hardcoding a list of files here is good to prevent the updated-version problem.  Maybe what we can do is to also scan for _BUNDLED_METADATA in files and if the two lists don't agree, make that an error as well.  That way failure to do the right thing in both places will lead to an error that can instruct users about what they forgot.

Reply:
Okay, added that additional check